File: content/5-memory/2-history-compaction/gemini/3-summarize.py
from google import genai
from google.genai import types

LLM_MODEL = "gemini-pro-latest"
SUMMARY_MODEL = "gemini-flash-latest"  # cheap model for summaries
client = genai.Client()

# summarization: compress old messages with a cheap model
# no built-in compaction API — call a fast model to summarize,
# then replace old messages with the summary
WINDOW = 10
history: list[types.Content] = []

def summarize_old(messages: list[types.Content]) -> list[types.Content]:
    if len(messages) <= 20:
        return messages  # short enough — no compression needed
    old, recent = messages[:-WINDOW], messages[-WINDOW:]
    summary = client.models.generate_content(
        model=SUMMARY_MODEL,
        contents=[
            types.Content(role="user", parts=[types.Part(
                text=f"Summarize this conversation in 2-3 sentences. "
                f"Preserve key facts and decisions.\n\n{old}",
            )]),
        ],
    )
    return [
        types.Content(role="user", parts=[types.Part(
            text=f"[Summary of earlier conversation]: {summary.text}",
        )]),
        *recent,
    ]

def chat(message: str) -> str:
    history.append(types.Content(role="user", parts=[types.Part(text=message)]))
    window = summarize_old(history)
    response = client.models.generate_content(
        model=LLM_MODEL, contents=window,
    )
    history.append(types.Content(role="model", parts=[types.Part(text=response.text)]))
    return response.text

print(chat("What is the capital of France?"))
print(chat("What is its population?"))

File: content/5-memory/2-history-compaction/aisdk/3-summarize.ts
import { generateText, pruneMessages, type ModelMessage } from "ai";
import { openai } from "@ai-sdk/openai";

const LLM_MODEL = "gpt-5.4";
const SUMMARY_MODEL = "gpt-5-mini"; // cheap model for summaries

// summarization: prune tool/reasoning bloat, then compress old messages
// pruneMessages is built-in (strips tool calls and reasoning content)
// LLM summarization is manual — uses a cheap model for compression
const messages: ModelMessage[] = [];

async function compactHistory(msgs: ModelMessage[]): Promise<ModelMessage[]> {
  // step 1: strip old tool calls and reasoning — built-in, zero-cost
  const pruned = pruneMessages({
    messages: msgs,
    reasoning: "before-last-message",
    toolCalls: "before-last-2-messages",
  });
  if (pruned.length <= 20) return pruned;
  // step 2: still too long — summarize older messages with a cheap model
  const old = pruned.slice(0, -10);
  const recent = pruned.slice(-10);
  const { text } = await generateText({
    model: openai(SUMMARY_MODEL),
    messages: [
      {
        role: "user",
        content:
          "Summarize this conversation in 2-3 sentences. " +
          `Preserve key facts and decisions.\n\n${JSON.stringify(old)}`,
      },
    ],
  });
  return [{ role: "assistant", content: `[Summary]: ${text}` }, ...recent];
}

async function chat(message: string): Promise<string> {
  messages.push({ role: "user", content: message });
  const compacted = await compactHistory(messages);
  const result = await generateText({
    model: openai(LLM_MODEL),
    messages: compacted,
  });
  messages.push({ role: "assistant", content: result.text });
  return result.text;
}

console.log(await chat("What is the capital of France?"));
console.log(await chat("What is its population?"));
